feat(http): adding a skeleton for a new mock http source helper

jarlah opened the pull request:
As suggested in issue #567 there is a need for mocking the http source.

Widdershin commented:
@jarlah very nice start. The `.select()` implementation looks good.

Do you want to have a go at adding some tests next?

Have a try and I can give you some examples if you get stuck.

Also seems like we need some way of specifying responses that don't have a category, maybe we could have a special key like `:root` at the top level. Not sure what the best way to handle that is, any thoughts?

jarlah commented:
@Widdershin yes, you might be right about the :root suggestion. But given the dynamic nature of type signatures in typescript (and flow also), its possible to say that MockConfig is either X, Y or Z. The problem with this of course, is that you need to do runtime type checks to find out what the heck you are working with. But, I am actua|lly more worried about the api and usage, so I would suggest saying ```type MockConfig = Responses | { [categoryName?]: Responses | MockConfig }``` but i am new here so I am not sure if this is something one would let slip through ;)

jarlah commented:
@Widdershin btw i made a test for empty stream. But I need some input on how to make a test for some "actual" response streams. Because the stream that is passed in from the test should match the type signature ```Stream<MemoryStream<Response> & ResponseStream>```

jarlah commented:
```/bin/sh: 3: Syntax error: "done" unexpected (expecting "do")```
cant see that I have introduced something that could cause this to fail travis

Widdershin commented:
I think the build failure is unrelated, seems to be an issue from staltz moving the config over to makefiles. I think he's on break for Easter, I'm sure he'll sort it out when he's back online.

I see the dilemma around the stream passed from the test matching `Stream<MemoryStream<Response> & ResponseStream>`. Personally it seems like it would be unreasonable to ask the user to specify all of the attributes involved in a `SuperAgentResponse`.

I think we can follow the precedent from `mockDOMSource`, which is to be permissive about what is actually in the streams. We are not forced to ensure that it is a `Stream<Event>`, it just has to be a stream.

So maybe we should try `export type Responses = Stream<any>` or `export type Responses = FantasyObservable`. We also want the user to optionally be able to set the request, so maybe make a type that is `Stream<any>` + optional request.

With those more permissive types, we can write a test like this:

```js
  it('supports categorized streams', (done) => {
    const response : any = {
      text: 'Hello world'
    }

    const response$ = xs.of(response);

    const userEvents = mockHTTPSource({
      'foo': xs.of(response$)
    });

    userEvents.select('foo').flatten().take(1).subscribe({
      next: (response: any) => {
        assert.equal(response.text, 'Hello world');
      },
      error: (err: any) => done(err),
      complete: () => done(),
    });
  });
```

Do you reckon that's a viable approach?

Widdershin commented on http/test/node/mock-http-source.ts:
```diff
@@ -0,0 +1,21 @@
+import 'mocha';
+import * as assert from 'assert';
+import { mockHTTPSource } from '../../lib/index';
```
We probably want to change this to import from `../../src/index`, it should be okay since mocha is configured to compile any typescript files it finds as it goes.

Widdershin commented:
Fyi @jarlah I fixed CI so if you pull in master you can run the tests on Travis now.

jarlah commented:
@Widdershin thnx. I'll dive into it the coming days. If it hadn't been for all this work related stuff... :laughing:

jarlah commented:
@Widdershin I have now updated my branch and added your exampel test, and changed  the Respones type to be more open. We can disccuss your point about specifying request

Widdershin commented:
@jarlah sorry for my lack of followup on this, I must have seen your comment and then forgot about it.

This implementation is looking good. I had a thought about how we could possibly get the types all nice as well, let me know what you think:

We could create helpers to create `Response` and `Request` objects.

Change this:
```diff
-    const response: any = {
-      text: 'Hello world',
-    };
+    const response = mockResponse({
+      text: 'Hello world',
+    });
```

And for requests:

```js
const request = mockRequest({
  url: '/anything',
  method: 'POST'
}

const responseToRequest = mockResponse({
  data: {},
  request
}
```

jarlah commented:
Someone else needs to pick this up. Need to get comfortable with cyclejs before go headlong into opensource pull requests like this one.

Widdershin commented:
I'll pick this up.
